activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -203,14 +287,18 @@ def internal_metadata # :nodoc:
       #
       # #connection can be called any number of times; the connection is
       # held in a cache keyed by a thread.
-      def connection
-        @thread_cached_conns[ActiveSupport::IsolatedExecutionState.context] ||= checkout
+      def lease_connection
+        lease = connection_lease
+        lease.sticky = true
+        lease.connection ||= checkout
       end
 
+      alias_method :connection, :lease_connection # TODO: deprecate
```

Comment:
This require changing a lot of tests so I'll do it in a followup.

Note: this is `ConnectionPool#connection`, not `AR::Base#connection`, the later is way too common to be realistically deprecated in the short term.